clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -393,25 +403,41 @@ boolean joinGroupIfNeeded(final Timer timer) {
                 // need to set the flag before calling onJoinPrepare since the user callback may throw
                 // exception, in which case upon retry we should not retry onJoinPrepare either.
                 needsJoinPrepare = false;
-                onJoinPrepare(generation.generationId, generation.memberId);
+                onJoinPrepare(gen.generationId, gen.memberId);
             }
 
-            final RequestFuture<ByteBuffer> future = initiateJoinGroup();
+            final RequestFuture<ByteBuffer> future = initiateJoinGroup(gen);
             client.poll(future, timer);
             if (!future.isDone()) {
                 // we ran out of time
                 return false;
             }
 
             if (future.succeeded()) {
-                // Duplicate the buffer in case `onJoinComplete` does not complete and needs to be retried.
-                ByteBuffer memberAssignment = future.value().duplicate();
-                onJoinComplete(generation.generationId, generation.memberId, generation.protocol, memberAssignment);
+                // Generation data maybe concurrently cleared by Heartbeat thread.
+                // Can't use synchronized for {@code onJoinComplete}, because it can be long enough
+                // and  shouldn't block hearbeat thread.
+                // See {@link PlaintextConsumerTest#testMaxPollIntervalMsDelayInAssignment
+                synchronized (this) {
+                    gen = this.generation;
```

Comment:
The `gen` object can still be an UNKNOWN generation since `this.generation` can be reset before the synchronization block, so just synchronize on fetching it would not help much I think.

On the other hand, I think I agree that synchronizing the `onJoinPrepare/Complete` would not be a good solution after second thought since they includes user-callback which can potentially take very long time. 
AND if the generation has actually be reset due to heartbeat failure or pro-active leaving group, it would be later notified by the coordinator that it is no longer in the group and re-join later anyways, so this race condition is actually "harmless".

What we really need to fix, is to make sure we did not pass in an UNKNOWN generation to the `onJoinComplete` below. So I'd propose we just copy the generation here atomically (not a reference assignment, but a hard copy) within this synchronization block, and then do the check below as is. This hard copy would not be modified by the other thread.